### PR TITLE
Discontinue Tribe Incentives

### PR DESCRIPTION
**Pull request by thomas-waite**
# Discontinue TRIBE incentives
Enact the passed TribalCouncil proposal to discontinue Tribe incentives. 

Specifically this:
- Unlocks all pools to allow principle to be withdrawn now that rewards have ended
- Sets all pool APs to zero, apart from pool 3 which has it's AP set 1 - to avoid potential divide by zero errors
- Sets the tribe block reward to 1e6, effectively 0. This globally discontinues rewards
- Removes the CREAM deposit from the CR

## Items for later
1. Harvest the staking token wrappers
2. Calculate how much TRIBE is owed out to users
3. Withdraw excess TRIBE from the TribalChief back to Treasury

### CODE REVIEW

**Review by ElliotFriedman — DISMISSED** [review]
LGTM!

**Comment by eswak** [comment]
Update of https://github.com/fei-protocol/fei-protocol-core/blob/develop/protocol-configuration/tribalchief.ts not needed ? I'm surprised

**Comment by thomas-waite** [comment]
> Update of https://github.com/fei-protocol/fei-protocol-core/blob/develop/protocol-configuration/tribalchief.ts not needed ? I'm surprised

Simplest solution is to just tell the TribalChief that new block rewards are 0. Alternative approach would have been to set the APs of all pools to 0

File: proposals/description/end-tribe-incentives.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+const end_tribe_incentives: ProposalDescription = {
+  // Pool ID is where the pool is in the array
+  title: 'TIP-109: Discontinue Tribe Incentives',
+  commands: [
+    {
+      target: 'tribalChief',
+      values: '0',
+      method: 'massUpdatePools(uint256[])',
+      arguments: [['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17']],
```
Comment (Joeysantoro) [inline proposals/description/end-tribe-incentives.ts]: you could skip all the pools with 0 AP

File: proposals/description/end_tribe_incentives.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+      description: 'Set Pool 3 rewards to 1 to allow principle withdrawal and do not overwrite or change the rewarder'
+    },
+    {
+      target: 'tribalChief',
+      values: '0',
+      method: 'set(uint256,uint120,address,bool)',
+      arguments: ['4', '0', '0x0000000000000000000000000000000000000000', false],
+      description: 'Set Pool 4 rewards to 0 to allow principle withdrawal and do not overwrite or change the rewarder'
+    },
+    {
```
Comment (thomas-waite) [inline proposals/description/end_tribe_incentives.ts:156]: Could skip those that already have AP = 0, but feel doing it en-masse instead is fail safe. No chance of making a mistake

File: proposals/description/end_tribe_incentives.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    /////////////  Replicate onlyGovernor behaviour or resetRewards()
+    ////  Unlock pools all pools to allow principle withdrawal
+    {
+      target: 'tribalChief',
+      values: '0',
+      method: 'unlockPool(uint256)',
+      arguments: ['0'],
+      description: 'Unlock pool 0'
+    },
```
Comment (Joeysantoro) [inline proposals/description/end_tribe_incentives.ts]: Pools were already unlocked, I'd delete this section

File: proposals/description/end_tribe_incentives.ts (hunk 2 above)
Comment (Joeysantoro) [reply proposals/description/end_tribe_incentives.ts:156]: I would leave exactly one pool with nonzero AP to be safe. Probably pool 8 but maybe the FEI-TRIBE LP

Comment (thomas-waite) [reply proposals/description/end_tribe_incentives.ts:156]: Have done, pool 3 (Fei-Rari, pool 8). In the block above this, set the AP point to 1

File: proposals/description/end_tribe_incentives.ts (hunk 3 above)
Comment (thomas-waite) [reply proposals/description/end_tribe_incentives.ts]: Etherscan reports as locked, https://etherscan.io/address/0x9e1076cC0d19F9B0b8019F384B0a29E48Ee46f7f#readProxyContract , `poolInfo`[1] states `unlocked = false`

Comment (thomas-waite) [reply proposals/description/end_tribe_incentives.ts]: Update: Removed as discussed

File: proposals/dao/end_tribe_incentives.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    await rewardsDistributorAdmin.hasRole(
+      ethers.utils.id('AUTO_REWARDS_DISTRIBUTOR_ROLE'),
+      addresses.fei3CrvAutoRewardsDistributor
+    )
+  ).to.equal(false);
+
+  // 4. Verify total AP points is 1
+  expect(await tribalChief.totalAllocPoint()).to.be.equal(NEW_TOTAL_ALLOC_POINTS.toString());
+
+  // 5. Validate that can harvest Tribe rewards
```
Comment (thomas-waite) [inline proposals/dao/end_tribe_incentives.ts]: TODO: Flesh out these into substantial e2e tests

File: proposals/description/end_tribe_incentives.ts (hunk 2 above)
Comment (thomas-waite) [reply proposals/description/end_tribe_incentives.ts:156]: Note: Updated to Fei-Tribe LP (pool id 0)

File: proposals/dao/end_tribe_incentives.ts (hunk 4 above)
Comment (thomas-waite) [reply proposals/dao/end_tribe_incentives.ts]: Note: Done in `tribeIncentivesEnd.ts`

File: test/integration/tests/staking.ts [hunk 5]
```diff
@@ -201,6 +201,8 @@ describe('e2e-staking', function () {
       await tribalChief.connect(daoSigner).set(16, 500, ethers.constants.AddressZero, false);
       await tribalChief.connect(daoSigner).set(17, 250, ethers.constants.AddressZero, false);
 
+      // TOOD: Grant various autoRewardDistributors the relevant role
```
Comment (Joeysantoro) [inline test/integration/tests/staking.ts]: should this test not just be deleted?

File: test/integration/tests/tribeIncentivesEnd.ts [hunk 6]
```diff
@@ -54,33 +56,43 @@ describe('e2e-end-tribe-incentives', function () {
...
   it('should NOT be able to harvest future TRIBE rewards from an LP pool', async () => {
-    // Fast forward time by a few blocks
-    // Try to withdraw
-    // Should withdraw basically zero
+    // Harvest, to zero out already earned rewards
+    const firstHarvestInitial = await tribe.balanceOf(receiver);
+    const stakerInPoolSigner = await getImpersonatedSigner(curvePoolStaker);
+    await tribalChief.connect(stakerInPoolSigner).harvest(curvePoolId, receiver);
+    const firstHarvestFinal = await tribe.balanceOf(receiver);
+    expect(firstHarvestFinal.sub(firstHarvestInitial)).to.be.bignumber.greaterThan(toBN(1));
```
Comment (Joeysantoro) [inline test/integration/tests/tribeIncentivesEnd.ts:84]: this test is brittle, won't it fail after the real onchain harvest?

File: test/integration/tests/staking.ts (hunk 5 above)
Comment (thomas-waite) [reply test/integration/tests/staking.ts]: Eventually I'd agree. I will probably delete it as part of the incentives cleanup PR - i.e. we pull the remaining TRIBE off the TribalChief and Votium contract. Until then, wanted to initialise these tests with expected state so they pass

File: proposals/description/end_tribe_incentives.ts [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+      method: 'set(uint256,uint120,address,bool)',
+      arguments: ['2', '0', '0x0000000000000000000000000000000000000000', false],
+      description: 'Set Pool 2 rewards to 0 and do not overwrite or change the rewarder'
+    },
+    {
+      target: 'tribalChief',
+      values: '0',
+      method: 'set(uint256,uint120,address,bool)',
+      arguments: ['3', '0', '0x0000000000000000000000000000000000000000', false],
+      description: 'Set Pool 3 rewards to 1 and do not overwrite or change the rewarder'
```
Comment (Joeysantoro) [inline proposals/description/end_tribe_incentives.ts]: nit: "set rewards to 0"

Comment (thomas-waite) [reply proposals/description/end_tribe_incentives.ts]: done

File: test/integration/tests/tribeIncentivesEnd.ts (hunk 6 above)
Comment (thomas-waite) [reply test/integration/tests/tribeIncentivesEnd.ts:84]: Yep good point. I'll add a check to see if the user has any pending rewards to harvest. 

If they have pending rewards, then perform the harvest to zero it out. If not, then do not perform the first harvest

Comment (thomas-waite) [reply test/integration/tests/tribeIncentivesEnd.ts:84]: Update: Unless feel strongly, want to leave this as is. Will then properly fix if it becomes an issue in the future, or punt to the clean up PR

**Review by Joeysantoro — APPROVED** [review]

File: proposals/dao/end_tribe_incentives.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+  // 6. Verify an AutoRewardDistributor speed is set correctly
+  const d3Ctoken = await contracts.rariPool8Comptroller.cTokensByUnderlying(addresses.curveD3pool);
+
+  const d3RewardSpeed = await contracts.d3AutoRewardsDistributor.getNewRewardSpeed();
+  expect(d3RewardSpeed[1]).to.equal(false); // updateNeeded
+  expect(d3RewardSpeed[0]).to.equal(0); // newSpeed
+
+  const delegatorReportedSpeed = await contracts.rariRewardsDistributorDelegator.compSupplySpeeds(d3Ctoken);
+  expect(delegatorReportedSpeed).to.equal(0);
+};
```
Comment (ElliotFriedman) [inline proposals/dao/end_tribe_incentives.ts:143]: you should verify here that all AutoRewardsDistributors are set to 0 alloc points as well.

**Pull request merged**